oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+
+* ``CONTEXT_TYPE``: The data is related to all instances of this type of XBlock within a particular context.
+* ``CONTEXT_ALL``: The data is accessible to XBlocks of any type, within a particular context. (As with ``BlockScope.ALL``, XBlock authors who use such fields should be careful to use a namespace that will not conflict with ``BlockScope.CONTEXT_ALL`` fields defined by other XBlock authors.)
+
+A "context" is defined by the runtime but is usually a "course" or the nearest equivalent. In the Open edX platform, the context may be a course or a content library. Different runs of the same course would be considered different contexts and would not share data, though UserScope.NONE fields would be copied when creating a new course run. On other platforms, a "context" could be defined as a blog post or a textbook.
+
+New named scopes
+----------------
+
+As is the situtation today, XBlock runtimes would generally not be expected to implement every possible combination of ``BlockScope`` and ``UserScope``. Instead, XBlock runtimes are encouraged to implement these new named scopes, which would be defined in the XBlock python package:
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this course
+
+* ``Scope.context_shared_settings`` (UserScope.NONE, BlockScope.CONTEXT_ALL)
+
+  * Example: Course-specific defaults that can apply to many different blocks, such as "max_attempts" and "show_answer", which would be consumed by the Capa Problem XBlock, Drag and Drop problem XBlock, Problem Builder XBlock family, etc. (These are currently implemented via ``InheritanceMixin``, which is built-in to edx-platform and does not allow XBlock plugins to extend the list of inherited fields. However, a downside of making them a context-scoped field is that each block would require two field definitions: a ``Scope.settings`` field for the usage-specific value, and a ``Scope.context_shared_settings`` field for the default/fallback value. )
+  * Example: Submission blackout dates - don't allow users to submit answers during the specified time period.
+
+* ``Scope.context_preferences`` (UserScope.ONE, BlockScope.CONTEXT_TYPE)
+
+  * Example: Enable or disable notifications about new discussion posts in this course
+  * Example: Video playback speed for this course
+  * Example: An XBlock in which students "accumulate building blocks over the course of the semester" (`PLAT-325`_)
+
+* ``Scope.context_shared_preferences`` (UserScope.ONE, BlockScope.CONTEXT_ALL)
+
+  * Example: Allow users to self-assign themselves into groups, which affects multiple types of problem XBlock (ORA, Problem Builder)?
+  * This seems to be the least-compelling new Scope, and may be dropped from the proposal if compelling use cases cannot be found.
```

Comment:
Sure, I will drop the `context_shared_preferences` scope. We could always add it later if anyone had a use case.